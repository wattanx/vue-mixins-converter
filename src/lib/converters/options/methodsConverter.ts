import ts from 'typescript';
import {
  ConvertedExpression,
  getInitializerProps,
  getMethodExpression,
} from '../../helper';

export const methodsConverter = (
  node: ts.Node,
  sourceFile: ts.SourceFile
): ConvertedExpression[] => {
  return getInitializerProps(node)
    .map((prop) => {
      return getMethodExpression(prop, sourceFile);
    })
    .flat();
};
