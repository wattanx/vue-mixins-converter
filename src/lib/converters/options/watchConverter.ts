import ts from 'typescript';
import {
  ConvertedExpression,
  getInitializerProps,
  nonNull,
} from '../../helper';

export const watchConverter = (
  node: ts.Node,
  sourceFile: ts.SourceFile
): ConvertedExpression[] => {
  return getInitializerProps(node)
    .map((prop) => {
      if (ts.isMethodDeclaration(prop)) {
        const name = prop.name.getText(sourceFile);
        const parameters = prop.parameters
          .map((param) => param.getText(sourceFile))
          .join(',');
        const block = prop.body?.getText(sourceFile) || '{}';

        return {
          use: 'watch',
          expression: `watch(${name}, (${parameters}) => ${block})`,
        };
      } else if (ts.isPropertyAssignment(prop)) {
        if (!ts.isObjectLiteralExpression(prop.initializer)) return;

        const props = prop.initializer.properties.reduce(
          (acc: Record<string, ts.ObjectLiteralElementLike>, prop) => {
            const name = prop.name?.getText(sourceFile);
            if (name) acc[name] = prop;
            return acc;
          },
          {}
        );

        const { handler, immediate, deep } = props;
        if (!(handler && ts.isMethodDeclaration(handler))) return;

        const options = [immediate, deep].reduce(
          (acc: Record<string, any>, prop) => {
            if (prop && ts.isPropertyAssignment(prop)) {
              const name = prop.name?.getText(sourceFile);
              if (name) {
                acc[name] = prop.initializer.kind === ts.SyntaxKind.TrueKeyword;
              }
            }
            return acc;
          },
          {}
        );

        const name = prop.name.getText(sourceFile);
        const parameters = handler.parameters
          .map((param) => param.getText(sourceFile))
          .join(',');
        const block = handler.body?.getText(sourceFile) || '{}';

        return {
          use: 'watch',
          expression: `watch(${name}, (${parameters}) => ${block}, ${JSON.stringify(
            options
          )} )`,
        };
      }
    })
    .filter(nonNull);
};
