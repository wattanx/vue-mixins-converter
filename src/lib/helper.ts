import ts, { factory } from 'typescript';

export type ConvertedExpression = {
  expression: string;
  returnNames?: string[];
  use?: string;
};

export const lifecycleNameMap: Map<string, string | undefined> = new Map([
  ['beforeCreate', undefined],
  ['created', undefined],
  ['beforeMount', 'onBeforeMount'],
  ['mounted', 'onMounted'],
  ['beforeUpdate', 'onBeforeUpdate'],
  ['updated', 'onUpdated'],
  ['beforeUnmount', 'onBeforeUnmount'],
  ['beforeDestroy', 'onBeforeUnmount'],
  ['destroyed', 'onUnmounted'],
  ['errorCaptured', 'onErrorCaptured'],
  ['renderTracked', 'onRenderTracked'],
  ['renderTriggered', 'onRenderTriggered'],
]);

export const nonNull = <T>(item: T): item is NonNullable<T> => item != null;

export const getNodeByKind = (
  node: ts.Node,
  kind: ts.SyntaxKind
): ts.Node | undefined => {
  const find = (node: ts.Node): ts.Node | undefined => {
    return ts.forEachChild(node, (child) => {
      if (child.kind === kind) {
        return child;
      }
      return find(child);
    });
  };
  return find(node);
};

export const getInitializerProps = (
  node: ts.Node
): ts.ObjectLiteralElementLike[] => {
  if (!ts.isPropertyAssignment(node)) return [];
  if (!ts.isObjectLiteralExpression(node.initializer)) return [];
  return [...node.initializer.properties];
};

export const storePath = `store`;

export const getMethodExpression = (
  node: ts.Node,
  sourceFile: ts.SourceFile
): ConvertedExpression[] => {
  if (ts.isMethodDeclaration(node)) {
    const async = node.modifiers?.some(
      (mod) => mod.kind === ts.SyntaxKind.AsyncKeyword
    )
      ? 'async'
      : '';

    const name = node.name.getText(sourceFile);
    const type = node.type ? `:${node.type.getText(sourceFile)}` : '';
    const body = node.body?.getText(sourceFile) || '{}';
    const parameters = node.parameters
      .map((param) => param.getText(sourceFile))
      .join(',');
    const fn = `${async}(${parameters})${type} =>${body}`;

    if (lifecycleNameMap.has(name)) {
      const newLifecycleName = lifecycleNameMap.get(name);
      const immediate = newLifecycleName == null ? '()' : '';
      return [
        {
          use: newLifecycleName,
          expression: `${newLifecycleName ?? ''}(${fn})${immediate}`,
        },
      ];
    }
    return [
      {
        returnNames: [name],
        expression: `const ${name} = ${fn}`,
      },
    ];
  } else if (ts.isSpreadAssignment(node)) {
    // mapActions
    if (!ts.isCallExpression(node.expression)) return [];
    const { arguments: args, expression } = node.expression;
    if (!ts.isIdentifier(expression)) return [];
    const mapName = expression.text;
    const [namespace, mapArray] = args;
    if (!ts.isStringLiteral(namespace)) return [];
    if (!ts.isArrayLiteralExpression(mapArray)) return [];

    const namespaceText = namespace.text;
    const names = mapArray.elements as ts.NodeArray<ts.StringLiteral>;

    if (mapName === 'mapActions') {
      return names.map(({ text: name }) => {
        return {
          expression: `const ${name} = () => ${storePath}.dispatch('${namespaceText}/${name}')`,
          returnNames: [name],
        };
      });
    }
  }
  return [];
};

const contextProps = [
  'attrs',
  'slots',
  'parent',
  'root',
  'listeners',
  'refs',
  'emit',
];

export const replaceThisContext = (
  str: string,
  refNameMap: Map<string, true>
) => {
  return str
    .replace(/this\.\$(\w+)/g, (_, p1) => {
      if (contextProps.includes(p1)) return `ctx.${p1}`;
      return `ctx.root.$${p1}`;
    })
    .replace(/this\.([\w-]+)/g, (_, p1) => {
      return refNameMap.has(p1) ? `${p1}.value` : p1;
    });
};

export const getImportStatement = (
  setupProps: ConvertedExpression[],
  useNuxt?: boolean
) => {
  let usedFunctions = [
    'SetupContext',
    ...new Set(setupProps.map(({ use }) => use).filter(nonNull)),
  ];

  const moduleName = useNuxt ? 'nuxtjs' : 'vue';

  if (useNuxt) {
    usedFunctions = [...usedFunctions, 'useStore'];
  }

  const compositionApiImportStatements = ts.createSourceFile(
    '',
    `import { ${usedFunctions.join(
      ','
    )} } from '@${moduleName}/composition-api'`,
    ts.ScriptTarget.Latest
  ).statements;

  const vuexImportStatements = ts.createSourceFile(
    '',
    `import { useStore } from 'vuex';`,
    ts.ScriptTarget.Latest
  ).statements;

  return useNuxt
    ? compositionApiImportStatements
    : [...vuexImportStatements, ...compositionApiImportStatements];
};

export const getExportStatement = (
  setupProps: ConvertedExpression[],
  fileName: string
) => {
  const useStoreStatements = ts.createSourceFile(
    '',
    'const store = useStore()',
    ts.ScriptTarget.Latest
  ).statements;

  const props = [
    factory.createParameterDeclaration(
      undefined,
      undefined,
      undefined,
      factory.createIdentifier('ctx'),
      undefined,
      factory.createTypeReferenceNode(
        factory.createIdentifier('SetupContext'),
        undefined
      ),
      undefined
    ),
  ];

  return factory.createVariableStatement(
    [factory.createModifier(ts.SyntaxKind.ExportKeyword)],
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createIdentifier(fileName),
          undefined,
          undefined,
          factory.createArrowFunction(
            undefined,
            undefined,
            props,
            undefined,
            factory.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
            factory.createBlock(
              [...useStoreStatements, ...getSetupStatements(setupProps)],
              true
            )
          )
        ),
      ],
      ts.NodeFlags.Const
    )
  );
};

export const getSetupStatements = (setupProps: ConvertedExpression[]) => {
  // this.prop => prop.valueにする対象
  const refNameMap: Map<string, true> = new Map();
  setupProps.forEach(({ use, returnNames }) => {
    if (
      returnNames != null &&
      use != null &&
      /^(toRefs|ref|computed)$/.test(use)
    ) {
      returnNames.forEach((returnName) => {
        refNameMap.set(returnName, true);
      });
    }
  });

  const returnPropsStatement = `return {${setupProps
    .filter((prop) => prop.use !== 'toRefs') // ignore spread props
    .map(({ returnNames }) => returnNames)
    .filter(nonNull)
    .flat()
    .join(',')}}`;

  return [...setupProps, { expression: returnPropsStatement }]
    .map(
      ({ expression }) =>
        ts.createSourceFile(
          '',
          replaceThisContext(expression, refNameMap),
          ts.ScriptTarget.Latest
        ).statements
    )
    .flat();
};
