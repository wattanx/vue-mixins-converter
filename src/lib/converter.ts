import ts from 'typescript';
import { getNodeByKind } from './helper';
import { convertOptionsApi } from './converters/optionsApiConverter';

export type ConverterOptions = {
  input: string;
  fileName: string;
  useNuxt?: boolean;
};

export const convertSrc = ({
  input,
  fileName,
  useNuxt,
}: ConverterOptions): string => {
  const scriptContent = input;

  const sourceFile = ts.createSourceFile(
    'src.tsx',
    scriptContent,
    ts.ScriptTarget.Latest
  );

  const exportAssignNode = getNodeByKind(
    sourceFile,
    ts.SyntaxKind.ExportAssignment
  );
  if (exportAssignNode) {
    // optionsAPI
    return convertOptionsApi(sourceFile, fileName, useNuxt);
  }

  throw new Error('no convert target');
};
