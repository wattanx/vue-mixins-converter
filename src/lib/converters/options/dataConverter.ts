import ts from 'typescript';
import { ConvertedExpression, getNodeByKind } from '../../helper';

export const dataConverter = (
  node: ts.Node,
  sourceFile: ts.SourceFile
): ConvertedExpression[] => {
  const objNode = getNodeByKind(node, ts.SyntaxKind.ObjectLiteralExpression);

  if (!(objNode && ts.isObjectLiteralExpression(objNode))) return [];
  return objNode.properties
    .map((prop) => {
      if (!ts.isPropertyAssignment(prop)) return;
      const name = prop.name.getText(sourceFile);
      const text = prop.initializer.getText(sourceFile);
      return {
        use: 'ref',
        expression: `const ${name} = ref(${text})`,
        returnNames: [name],
      };
    })
    .filter((item): item is NonNullable<typeof item> => item != null);
};
