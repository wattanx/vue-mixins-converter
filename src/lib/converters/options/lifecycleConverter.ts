import ts from 'typescript';
import { ConvertedExpression, getMethodExpression } from '../../helper';

export const lifecycleConverter = (
  node: ts.Node,
  sourceFile: ts.SourceFile
): ConvertedExpression[] => {
  return getMethodExpression(node, sourceFile);
};
