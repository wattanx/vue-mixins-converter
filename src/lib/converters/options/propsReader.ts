import ts from 'typescript';
import { nonNull } from '../../helper';

export const propReader = (
  node: ts.Node,
  sourceFile: ts.SourceFile
): string[] => {
  if (!ts.isPropertyAssignment(node)) return [];

  if (ts.isObjectLiteralExpression(node.initializer)) {
    return node.initializer.properties
      .map((prop) => {
        if (!ts.isPropertyAssignment(prop)) return null;
        return prop.name.getText(sourceFile);
      })
      .filter(nonNull);
  } else if (ts.isArrayLiteralExpression(node.initializer)) {
    return node.initializer.elements
      .map((el) => {
        if (ts.isStringLiteral(el)) return el.text;
      })
      .filter(nonNull);
  }
  return [];
};
