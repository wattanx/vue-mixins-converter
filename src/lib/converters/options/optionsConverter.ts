import ts from 'typescript';
import {
  ConvertedExpression,
  getNodeByKind,
  lifecycleNameMap,
} from '../../helper';
import { computedConverter } from './computedConverter';
import { dataConverter } from './dataConverter';
import { lifecycleConverter } from './lifecycleConverter';
import { methodsConverter } from './methodsConverter';
import { watchConverter } from './watchConverter';
import { propReader } from './propsReader';

export const convertOptions = (sourceFile: ts.SourceFile) => {
  const exportAssignNode = getNodeByKind(
    sourceFile,
    ts.SyntaxKind.ExportAssignment
  );
  if (exportAssignNode) {
    const objectNode = getNodeByKind(
      exportAssignNode,
      ts.SyntaxKind.ObjectLiteralExpression
    );
    if (objectNode && ts.isObjectLiteralExpression(objectNode)) {
      return _convertOptions(objectNode, sourceFile);
    }
  }
  const classNode = getNodeByKind(sourceFile, ts.SyntaxKind.ClassDeclaration);
  if (classNode) {
    const decoratorNode = getNodeByKind(classNode, ts.SyntaxKind.Decorator);

    if (decoratorNode) {
      const objectNode = getNodeByKind(
        decoratorNode,
        ts.SyntaxKind.ObjectLiteralExpression
      );

      if (objectNode && ts.isObjectLiteralExpression(objectNode)) {
        return _convertOptions(objectNode, sourceFile);
      }
    }
  }

  return null;
};

const _convertOptions = (
  exportObject: ts.ObjectLiteralExpression,
  sourceFile: ts.SourceFile
) => {
  const otherProps: ts.ObjectLiteralElementLike[] = [];
  const dataProps: ConvertedExpression[] = [];
  const computedProps: ConvertedExpression[] = [];
  const methodsProps: ConvertedExpression[] = [];
  const watchProps: ConvertedExpression[] = [];
  const lifecycleProps: ConvertedExpression[] = [];
  const propNames: string[] = [];

  exportObject.properties.forEach((prop) => {
    const name = prop.name?.getText(sourceFile) || '';
    switch (true) {
      case name === 'data':
        dataProps.push(...dataConverter(prop, sourceFile));
        break;
      case name === 'computed':
        computedProps.push(...computedConverter(prop, sourceFile));
        break;
      case name === 'watch':
        watchProps.push(...watchConverter(prop, sourceFile));
        break;
      case name === 'methods':
        methodsProps.push(...methodsConverter(prop, sourceFile));
        break;
      case lifecycleNameMap.has(name):
        lifecycleProps.push(...lifecycleConverter(prop, sourceFile));
        break;

      default:
        if (name === 'props') {
          propNames.push(...propReader(prop, sourceFile));
        }

        // 該当しないものはそのままにする
        otherProps.push(prop);
        break;
    }
  });

  const propsRefProps: ConvertedExpression[] =
    propNames.length === 0
      ? []
      : [
          {
            use: 'toRefs',
            expression: `const { ${propNames.join(',')} } = toRefs(props)`,
            returnNames: propNames,
          },
        ];

  const setupProps: ConvertedExpression[] = [
    ...propsRefProps,
    ...dataProps,
    ...computedProps,
    ...methodsProps,
    ...watchProps,
    ...lifecycleProps,
  ];

  return {
    setupProps,
    propNames,
    otherProps,
  };
};
