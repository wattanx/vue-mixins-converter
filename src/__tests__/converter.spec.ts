import { getInitializerProps } from './../lib/helper';
import { convertOptions } from './../lib/converters/options/optionsConverter';
import { convertSrc } from './../lib/converter';
import { it, expect } from 'vitest';
import fs from 'fs';
import { ts, Project, ScriptTarget, SyntaxKind, Node } from 'ts-morph';

const project = new Project({
  compilerOptions: { target: ScriptTarget.Latest },
});

it('convertSrc', () => {
  const mixin = fs.readFileSync('src/mixins/MixinSample.js').toString('utf-8');

  const converted = convertSrc({ input: mixin, fileName: 'useMixins' });

  expect(converted).toMatchSnapshot();
});

it('getInitializerProps', () => {
  const mixin = fs.readFileSync('src/mixins/MixinSample.js').toString('utf-8');
  const sourceFile = project.createSourceFile('src.tsx', mixin);
  const exportAssignNode = sourceFile.getFirstChildByKind(
    SyntaxKind.ExportAssignment
  );
  const objectNode = exportAssignNode?.getFirstChildByKindOrThrow(
    SyntaxKind.ObjectLiteralExpression
  );
  const props = objectNode?.getProperty('props');
  const data = objectNode?.getProperty('data');
  const computed = objectNode?.getProperty('computed');

  const methods = objectNode
    ?.getPropertyOrThrow('methods')
    .getFirstChildByKind(SyntaxKind.ObjectLiteralExpression);
  methods?.forEachChild((node) => {
    console.log(node.getKindName());
  });
  // methods.map((x) => {
  //   const spread = x.getProperty(
  //     (p) => p.getKind() === SyntaxKind.SpreadAssignment
  //   );
  // });
});
