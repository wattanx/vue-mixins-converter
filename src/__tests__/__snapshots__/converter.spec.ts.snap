// Vitest Snapshot v1

exports[`convertSrc 1`] = `
"import { useStore } from \\"vuex\\";
import { toRefs, ref, computed } from \\"@vue/composition-api\\";
export const useMixins = (props, ctx) => {
    const store = useStore();
    const { age } = toRefs(props);
    const firstName = ref(\\"first\\");
    const lastName = ref(\\"last\\");
    const name = computed(() => store.state.user.name);
    const age = computed(() => store.state.user.age);
    const name = computed(() => store.getters[\\"user/name\\"]);
    const age = computed(() => store.getters[\\"user/age\\"]);
    const fullName = computed(() => {
        return firstName.value + lastName.value;
    });
    const setUser = () => store.dispatch(\\"user/setUser\\");
    return { firstName, lastName, name, age, name, age, fullName, setUser };
};
"
`;
